src/osdc/Objecter.cc

```diff
@@ -3152,19 +3200,10 @@ void Objecter::list_nobjects(NListContext *list_context, Context *onfinish)
     return;
   }
 
-  rwlock.get_read();
-  const pg_pool_t *pool = osdmap->get_pg_pool(list_context->pool_id);
-  int pg_num = pool->get_pg_num();
-  rwlock.unlock();
-
-  if (list_context->starting_pg_num == 0) {     // there can't be zero pgs!
-    list_context->starting_pg_num = pg_num;
-    ldout(cct, 20) << pg_num << " placement groups" << dendl;
-  }
   if (list_context->starting_pg_num != pg_num) {
```

Comment:
This check needs to be done earlier, so that a newly computed pg_min/pg_max can be determined.  After a split a sharded ls could return completely different results to each worker/client.  The current_pg_epoch is advanced when a split is detected, but unless that caller is watching, it won't know that the listing had to restart.  Unfortunately, a user of NObjectIterator mechanism doesn't have access to this information.  It would seem to me that a client of this new feature might be sensitive to a split causing each objects to show up in multiple different shards.